go/decompress/decompress.go

```diff
@@ -0,0 +1,205 @@
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func ReadLineBgzip(r *bgzf.Reader) ([]byte, error) {
+	tx := r.Begin()
+	data, err := _readLineBgzip(r)
+	tx.End()
+	return data, err
+}
+
+func ReadLinesBgzip(r *bgzf.Reader) ([]byte, error) {
+	buf := make([]byte, 64*1024*8) // 8 bgzip blocks at a time
+
+	tx := r.Begin()
+	defer tx.End()
+
+	bytesRead, err := r.Read(buf)
+
+	if bytesRead == 0 {
+		return nil, err
+	}
+
+	if buf[bytesRead-1] != '\n' {
+		remainder, err := _readLineBgzip(r)
+		return append(buf[:bytesRead], remainder...), err
+	}
+	// last byte is newline
+	return buf[:bytesRead-1], err
+}
+
+func ReadLine(r *bufio.Reader) ([]byte, error) {
+	var (
+		data []byte
+		b    byte
+		err  error
+	)
+	for {
+		b, err = r.ReadByte()
+		if err != nil {
+			break
+		}
+		if b == '\n' {
+			break
+		}
+		data = append(data, b)
+	}
+	return data, err
+}
+
+func ReadLines(r *bufio.Reader) ([]byte, error) {
+	buf := make([]byte, 64*1024*8) // 8 bgzip blocks at a time
+
+	bytesRead, err := r.Read(buf)
+
+	if bytesRead == 0 {
+		return nil, err
+	}
+
+	if buf[bytesRead-1] != '\n' {
+		remainder, err := ReadLine(r)
+		return append(buf[:bytesRead], remainder...), err
+	}
+	// last byte is newline
+	return buf[:bytesRead-1], err
+}
+
+func (r *BzfBystroReader) ReadLines() ([]byte, error) {
+	return ReadLinesBgzip(r.Reader)
+}
+
+func (r *BzfBystroReader) ReadLine() ([]byte, error) {
+	return ReadLineBgzip(r.Reader)
+}
+
+func (r *BufioBystroReader) ReadLines() ([]byte, error) {
+	return ReadLines(r.Reader)
+}
+func (r *BufioBystroReader) ReadLine() ([]byte, error) {
+	return r.Reader.ReadBytes('\n')
+}
+
+func GetHeaderPaths(b BystroReader) ([][]string, []string) {
+	line, err := b.ReadLine()
+	if err != nil {
+		log.Fatalf("Error reading header line due to: [%s]\n", err)
+	}
+
+	headers := strings.Fields(string(line))
+
+	headerPaths := [][]string{}
+
+	for _, header := range headers {
+		path := strings.Split(header, ".")
+
+		headerPaths = append(headerPaths, path)
+	}
+
+	return headerPaths, headers
+}
+
+func GetAnnotationFhFromTarArchive(archive *os.File) (BystroReader, fs.FileInfo, error) {
+	tarReader, fileStats, err := _getAnnotationFhFromTarArchive(archive)
+
+	if err != nil {
+		return nil, nil, err
+	}
+
+	b, err := bgzf.NewReader(tarReader, 0)
+	if err != nil {
+		archive.Seek(0, 0)
+
+		tarReader, fileStats, err := _getAnnotationFhFromTarArchive(archive)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		b, err := gzip.NewReader(tarReader)
+
+		if err != nil {
+			return nil, nil, err
+		}
+
+		bufioReader := bufio.NewReader(b)
+
+		return &BufioBystroReader{Reader: bufioReader}, fileStats, nil
+	}
+
+	return &BzfBystroReader{Reader: b}, fileStats, err
+}
+
+func _getAnnotationFhFromTarArchive(archive *os.File) (*tar.Reader, fs.FileInfo, error) {
```

Comment:
Same issue with underscore as the first comment. Also, I'd suggest slightly refactoring to filehandleFromTarArchive so the comment would be // filehandleFromTarArchive takes a pointer to a tar archive and a filename and returns a desired filehandle from the tar archive.